File: 2025-2/sesion-de-clase-18/SoftProg/SoftProgPersistencia/src/main/java/pe/edu/pucp/inf30/softprog/daoimpl/rrhh/CuentaUsuarioDAOImpl.java
package pe.edu.pucp.inf30.softprog.daoimpl.rrhh;

import java.sql.CallableStatement;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Types;
import pe.edu.pucp.inf30.softprog.dao.rrhh.CuentaUsuarioDAO;
import pe.edu.pucp.inf30.softprog.daoimpl.TransaccionalBaseDAO;
import pe.edu.pucp.inf30.softprog.modelo.rrhh.CuentaUsuario;

/**
 *
 * @author eric
 */
public class CuentaUsuarioDAOImpl extends TransaccionalBaseDAO<CuentaUsuario> 
        implements CuentaUsuarioDAO {
    
    @Override
    protected PreparedStatement comandoCrear(Connection conn, 
            CuentaUsuario modelo) throws SQLException {
        
        String sql = "{call insertarCuentaUsuario(?, ?, ?, ?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setString("p_userName", modelo.getUserName());
        cmd.setString("p_password", modelo.getPassword());
        cmd.setBoolean("p_activo", modelo.isActivo());
        cmd.registerOutParameter("p_id", Types.INTEGER);
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoActualizar(Connection conn, 
            CuentaUsuario modelo) throws SQLException {
        
        String sql = "{call modificarCuentaUsuario(?, ?, ?, ?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setString("p_userName", modelo.getUserName());
        cmd.setString("p_password", modelo.getPassword());
        cmd.setBoolean("p_activo", modelo.isActivo());
        cmd.setInt("p_id", modelo.getId());
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoEliminar(Connection conn, 
            Integer id) throws SQLException {
        
        String sql = "{call eliminarCuentaUsuario(?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_id", id);
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeer(Connection conn, 
            Integer id) throws SQLException {
        
        String sql = "{call buscarCuentaUsuarioPorId(?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_id", id);
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeerTodos(Connection conn) 
            throws SQLException {
        
        String sql = "{call listarCuentaUsuarios()}";
                
        CallableStatement cmd = conn.prepareCall(sql);
        
        return cmd;
    }

    @Override
    protected CuentaUsuario mapearModelo(ResultSet rs) throws SQLException {
        CuentaUsuario cuentaUsuario = new CuentaUsuario();
        cuentaUsuario.setId(rs.getInt("id"));
        cuentaUsuario.setUserName(rs.getString("userName"));
        cuentaUsuario.setPassword(rs.getString("password"));
        cuentaUsuario.setActivo(rs.getBoolean("activo"));

        return cuentaUsuario;
    }    

    protected PreparedStatement comandoLogin(Connection conn, String userName, 
            String password) throws SQLException {
        String sql = "{call loginUsuario(?, ?, ?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setString("p_username", userName);
        cmd.setString("p_password", password);
        cmd.registerOutParameter("p_valido", Types.BOOLEAN);
        
        return cmd;
    }
    
    @Override
    public boolean login(String username, String password) {
        return ejecutarComando(conn -> {
            try (PreparedStatement cmd = this.comandoLogin(conn, username, password)) {
                if (cmd instanceof CallableStatement callableCmd) {
                    callableCmd.execute();
                    boolean valido = callableCmd.getBoolean("p_valido");
                    
                    if (!valido) {
                        System.err.println("No se encontro el registro con "
                            + "username: " + username + ", password");
                    }
                    return valido;
                }
                return false;
            }
        });
    }
}

File: 2025-2/sesion-de-clase-21/.net/SoftProgPersistencia/DAOImpl/RRHH/CuentaUsuarioDAOImpl.cs
using PUCP.SoftProg.Db;
using PUCP.SoftProg.Modelo.RRHH;
using PUCP.SoftProg.Persistencia.DAO.RRHH;
using System;
using System.Data;
using System.Data.Common;
using System.Net;

namespace PUCP.SoftProg.Persistencia.DAOImpl.RRHH {
    public class CuentaUsuarioDAOImpl : TransactionalBaseDAOImpl<CuentaUsuario>, ICuentaUsuarioDAO {
        protected override DbCommand ComandoCrear(DbConnection conn, CuentaUsuario cuentaUsuario) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "insertarCuentaUsuario";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_userName", DbType.String, cuentaUsuario.UserName);
            this.AgregarParametroEntrada(cmd, "@p_password", DbType.String, cuentaUsuario.Password);
            this.AgregarParametroEntrada(cmd, "@p_activo", DbType.Boolean, cuentaUsuario.IsActive);
            this.AgregarParametroSalida(cmd, "@p_id", DbType.Int32);

            return cmd;
        }

        protected override DbCommand ComandoActualizar(DbConnection conn, CuentaUsuario cuentaUsuario) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "modificarCuentaUsuario";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_userName", DbType.String, cuentaUsuario.UserName);
            this.AgregarParametroEntrada(cmd, "@p_password", DbType.String, cuentaUsuario.Password);
            this.AgregarParametroEntrada(cmd, "@p_activo", DbType.Boolean, cuentaUsuario.IsActive);
            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, cuentaUsuario.Id);

            return cmd;
        }

        protected override DbCommand ComandoEliminar(DbConnection conn, int id) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "eliminarCuentaUsuario";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand ComandoLeer(DbConnection conn, int id) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "buscarCuentaUsuarioPorId";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand ComandoLeerTodos(DbConnection conn) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "listarCuentaUsuarios";
            cmd.CommandType = CommandType.StoredProcedure;
            return cmd;
        }

        protected override CuentaUsuario MapearModelo(DbDataReader reader) {
            return new CuentaUsuario {
                Id = Convert.ToInt32(reader["id"]),
                UserName = Convert.ToString(reader["userName"]),
                Password = Convert.ToString(reader["password"]),
                IsActive = (bool)reader["activo"],
            };
        }

        protected DbCommand ComandoLogin(DbConnection conn, string userName, string password) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "loginUsuario";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_username", DbType.String, userName);
            this.AgregarParametroEntrada(cmd, "@p_password", DbType.String, password);
            this.AgregarParametroSalida(cmd, "@p_valido", DbType.Boolean);

            return cmd;
        }

        public bool Login(string username, string password) {
            return EjecutarComando(conn => {
                using (DbCommand cmd = this.ComandoLogin(conn, username, password)) {
                    cmd.ExecuteNonQuery();
                    return Convert.ToBoolean(cmd.Parameters["@p_valido"].Value);
                }
            });
        }
    }
}
